simpleviewinc/keg-hub
# feat: Added github actions for keg-hub

**Ticket**: [ZEN-392](https://jira.simpleviewtools.com/browse/ZEN-392)

[Semver-Status] minor

## Context

* We need to setup CI with github actions to automate common tasks


## Goal

* Setup some inital github actions for common tasks
* The following actions were added
  * **Build** - Runs on `push` to `master`
  * **Format** - Runs on `push` to `develop`
  * **PR-Changes** - Runs on **any** `pull_request`
  * **Tests** - Runs on **any** `pull_request`
* Added helper scripts for running the actions


## Updates
* `.github/workflows/*`
  * Github workflow files to run actions
* `scripts/ci/*`
  * Helper scripts run by the github actions


## Testing

**Setup**
* Fork the `keg-hub` repo into you're own account
* Pull the fork locally, or update your local repo to point to your fork
* **All the following tests should use your forked repo!**

**Test 1**
* After ensuring your git origin is poiting to your fork, switch to the `keg-github-actions` branch
* Next modify some files in different ways in a few different repos
  * For example, add a file to `Re-Theme`, and delete a file in `Keg-Core`
  * **IMPORTANT** - Do not modify all repos. We want to test that only the changed repos are effected
* Commit the changes you made, and push them to github.
  * Again make sure you push to **your own forked repo**
* After pushing to github, ensure no actions run

**Test 2**
* Create a pull request from the recently pushed branch ( that has your changes ) into **your forked version's** develop branch
* At this point, two of the github actions should run
  * `keg-test` and `keg-pr-changes`
  * Ensure both actions run and succeed
  * The `keg-test` workflow might fail depending on your changes
    * If it does, you've validated that it works properly ( good job )
    * Go revert the changes you made to brake the tests, and start over
    * 
  * The `keg-pr-changes`
    * This workflow shows a warning about invalid `inputs`
    * This is an issue with the action.yml file, that the creator needs to update
    * It's just a warning though, and should work as expected
* Assuming both workflows have succeeded
  * Go to the github action called `Run Repo Tests`
    * Toggle open the `Run Repo Tests` step
      * Ensure tests were only run for the repos that were changed
        * [Looks like this](http://snpy.in/5TUg5E)
  * Next, go to the PR's discussion tab
    * You should see a comment with a list of all changed files separated by repo
      * [Looks like this](http://snpy.in/67bObv)
    * Ensure only the repos that had changes are listed in the auto-generated comment


**Test 3**
* Merge the changes from you branch into the develop branch of **your fork**
  * **IMPORTANT** - Be sure it's your fork, and not the simpleviewinc one
* This should kick off the `keg-format` workflow
  * Go to the actions tab, and validate that the workflow ran successfully
  * [Looks like this](http://snpy.in/b5ORTB)
* Next goto the commit history for the develop branch
* Ensure a new commit was added with the formatted code
  * [Looks like this](http://snpy.in/Y9tMrL)

**Test 4**
* Finally, merge **your** develop branch into **your** the master branch
  * You'll probably have to do it from the command line
  * Make sure you pull down the changes to develop first
    * These changes are from the `keg-format` workflow of **Test 3**
* This should kick off the `keg-build` workflow
  * It will run tests, then build each one of the repos that has changed files
  * [Looks like this](http://snpy.in/a9u91S)
* It will then commit those build changes back into the master branch
  * Ensure a new commit was added with the built code
  * [Looks like this](http://snpy.in/VyudLV)
